config/repository.go

```diff
@@ -0,0 +1,271 @@
+	}
+}
+
+func (l *levelDBRepo) getTenantKey(id []byte) []byte {
+	return append([]byte(TenantPrefix), id...)
+}
+
+func (l *levelDBRepo) getConfigKey() []byte {
+	return []byte(ConfigPrefix)
+}
+
+// getModel returns a new instance of the type mt.
+func (l *levelDBRepo) getModel(mt string) (Model, error) {
+	tp, ok := l.models[mt]
+	if !ok {
+		return nil, fmt.Errorf("type %s not registered", mt)
+	}
+
+	return reflect.New(tp).Interface().(Model), nil
+}
+
+func (l *levelDBRepo) GetTenant(id []byte) (Model, error) {
+	key := l.getTenantKey(id)
+	return l.get(key)
+}
+
+func (l *levelDBRepo) GetConfig() (Model, error) {
+	key := l.getConfigKey()
+	return l.get(key)
+}
+
+func (l *levelDBRepo) parseModel(data []byte) (Model, error) {
+	v := new(value)
+	err := json.Unmarshal(data, v)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal value: %v", err)
+	}
+
+	nm, err := l.getModel(v.Type)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get model type: %v", err)
+	}
+
+	err = nm.FromJSON([]byte(v.Data))
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal to model: %v", err)
+	}
+
+	return nm, nil
+}
+
+// Get returns the model associated with ID
+func (l *levelDBRepo) get(id []byte) (Model, error) {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	data, err := l.db.Get(id, nil)
+	if err != nil {
+		return nil, fmt.Errorf("config missing: %v", err)
+	}
+
+	return l.parseModel(data)
+}
+
+// GetAllTenants iterates over all tenant entries in DB and returns a list of Models
+// If an error occur reading a tenant, throws a warning and continue
+func (l *levelDBRepo) GetAllTenants() ([]Model, error) {
+	var models []Model
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	iter := l.db.NewIterator(util.BytesPrefix([]byte(TenantPrefix)), nil)
+	for iter.Next() {
+		data := iter.Value()
+		model, err := l.parseModel(data)
+		if err != nil {
+			log.Warningf("Error parsing tenant: %v", err)
+			continue
+		}
+		models = append(models, model)
+	}
+	iter.Release()
+	return models, iter.Error()
+}
+
+// save stores the model.
+func (l *levelDBRepo) save(id []byte, model Model) error {
+	data, err := model.JSON()
+	if err != nil {
+		return fmt.Errorf("failed to marshall model: %v", err)
+	}
+
+	tp := getTypeIndirect(model.Type())
+	v := value{
+		Type: tp.String(),
+		Data: json.RawMessage(data),
+	}
+
+	data, err = json.Marshal(v)
+	if err != nil {
+		return fmt.Errorf("failed to marshall value: %v", err)
+	}
+
+	err = l.db.Put(id, data, nil)
+	if err != nil {
+		return fmt.Errorf("failed to save model to DB: %v", err)
+	}
+
+	return nil
+}
+
+// Exists returns true if the id exists.
+func (l *levelDBRepo) exists(id []byte) bool {
+	res, err := l.db.Has(id, nil)
+	// TODO check this
+	if err != nil {
+		return false
+	}
+
+	return res
+}
+
+// Create creates the tenant config model if not present in the DB.
+// should error out if the config exists.
+func (l *levelDBRepo) CreateTenant(id []byte, model Model) error {
```

Comment:
I would like to be more specific on names like `CreateTenantConfig` here as there is going to be another thing thats called `Tenant` in the system

Reply:
The Config abstraction is done in the Repository level, so config.Repository.CreateTenant() already has the `config` context. 
This might change a little after I am done with abstracting the DB backend btw, let's bring back this in the following PR